### PR TITLE
feat(packageresolution): add lockfileVersion 2+ packages support

### PR DESCRIPTION
**Pull request by anas10**
Resolves https://github.com/ds300/patch-package/issues/393

### CODE REVIEW

**Comment by julestruong** [comment]
good job :) when do you plan to merge this ?

**Comment by orta** [comment]
Cool - did you verify it works @julestruong?

@anas10  it'd be great to get an integration test for this, because I don't know enough about npm's lock files to know if it is correct: https://github.com/ds300/patch-package/tree/master/integration-tests (and to make sure no-one breaks it further down the line)

**Comment by julestruong** [comment]
I used patch package and yes it works

**Comment by anas10** [comment]
@orta Good shout, I've now added some integration tests for every version of lockfile.

**Comment by anas10** [comment]
Tests updated. I've made sure that they were not false positives.
Tests fail as expected without the code change and succeed consistently with the code change ✅

File: integration-tests/lockfile/lockfile.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# make sure errors stop the script
+set -e
+
+echo "add patch-package"
+npm i $1
+alias patch-package=./node_modules/.bin/patch-package
+
+function testLockFile() {
```
Comment (orta) [inline integration-tests/lockfile/lockfile.sh:8]: Perhaps this syntax is only available on your machine and now on CI?

Comment (anas10) [reply integration-tests/lockfile/lockfile.sh]: That's odd. This is a fairly standard syntax 🤔 
I've made a small change, can you trigger now to see. If not, I'll just repeat the test 3 times without the function.

**Comment by orta** [comment]
No idea _why_ - but I think we might need this on CI, https://github.com/grover/homebridge-dacp/issues/66#issuecomment-592986242

**Comment by iTonyYo** [comment]
Why not just judge by `lockfileVersion`?

**Comment by rokiyama** [comment]
@anas10 Just delete `#!/usr/bin/env bash` and CI will succeed. Alternatively, changing it to `#!/bin/sh` will also work.

This is because bash does not expand aliases when in non-interactive mode and shopt `expand_aliases` is not set.

succeed:

```sh
#!/bin/sh
alias my-command='echo hello'
my-command

# => hello
```

fail:

```sh
#!/bin/bash
alias my-command='echo hello'
my-command

# => line 3: my-command: command not found
```

**Comment by rokiyama** [comment]
Other problems:

While not critical, it might be a good idea to install `left-pad@1.3.0`.

Now lockfile v2 and v3 work. However, v1 still does not work.

`npm i --lockfile-version 1` produces error: `... /src/mdns.hpp:31:10: fatal error: dns_sd.h: No such file or directory.`

To fix this, adding `sudo apt install libavahi-compat-libdnssd-dev` to main.yaml results in another error: `... /src/mdns_utils.hpp:14:5: error: 'Handle' in namespace 'v8' does not name a template type`

I don't know how to resolve this. As a workaround, we can avoid the CI error by deleting the lockfile v1 test.

How I fixed it: https://github.com/Threestup/patch-package/pull/1/files

**Comment by rokiyama** [comment]
@orta Tests are updated, can you retrigger the tests?

**Comment by orta** [comment]
This PR looks good! 

I'm not at my main computer for 2 weeks, so I can't really handle the merge conflict for you - it should probably be automatic, it's changing a line next to one of your changes. If you get that fixed, I'll merge the PR and get a release sorted!

**Comment by anas10** [comment]
@orta I've resolved the conflicts. Thanks for checking

**Comment by orta** [comment]
Thanks everyone! Looks good to me

**Comment by ds300** [comment]
Released in [v6.5.1](https://github.com/ds300/patch-package/releases/tag/v6.5.1), thanks for the contribution 🎉 ❤️

**Pull request merged**